File: conformance/go/testee/recorded_wire.go
package main

// This is an acceptance witness, not the generated record/follow API. Its
// in-memory append store and subscriber writers are consumer compositions of
// Wire. Keeping them in the testee makes that boundary explicit.
import (
	"context"
	"encoding/json"
	"fmt"
	"sync"

	"github.com/Bitspark/nightseam/duplex/go"
)

type recordedEntry struct {
	path     []string
	message  duplex.Message
	sequence int
}

type recordedWire struct {
	mu        sync.Mutex
	entries   []recordedEntry
	followers map[*recordedFollower]bool
}

type recordedFollower struct {
	target duplex.Wire
	live   chan recordedEntry
	stop   chan struct{}
	done   chan struct{}
	sent   chan int
	paused chan struct{}
	resume chan struct{}
	once   sync.Once
}

func newRecordedWire() *recordedWire { return &recordedWire{followers: map[*recordedFollower]bool{}} }
func (w *recordedWire) head() int    { w.mu.Lock(); defer w.mu.Unlock(); return len(w.entries) }
func (w *recordedWire) Send(path []string, message duplex.Message) error {
	w.mu.Lock()
	entry := recordedEntry{append([]string{}, path...), message, len(w.entries) + 1}
	w.entries = append(w.entries, entry)
	for follower := range w.followers {
		select {
		case follower.live <- entry:
		default:
			delete(w.followers, follower)
			// Signal only. The subscriber's own writer closes its carrier, off
			// the producer's stack and outside append exclusion.
			follower.once.Do(func() { close(follower.stop) })
		}
	}
	w.mu.Unlock()
	return nil
}
func (*recordedWire) Receive([]string, duplex.Receiver) (func(), error) {
	return nil, duplex.ErrNoRoute
}
func (w *recordedWire) Close(duplex.Code, string) error {
	w.mu.Lock()
	for follower := range w.followers {
		follower.once.Do(func() { close(follower.stop) })
	}
	w.followers = map[*recordedFollower]bool{}
	w.mu.Unlock()
	return nil
}

func (w *recordedWire) attach(after int, target duplex.Wire, bound int, pause bool) (int, *recordedFollower) {
	f := &recordedFollower{target: target, live: make(chan recordedEntry, bound), stop: make(chan struct{}), done: make(chan struct{}), sent: make(chan int, 32), paused: make(chan struct{}), resume: make(chan struct{})}
	w.mu.Lock()
	head := len(w.entries)
	history := append([]recordedEntry{}, w.entries[after:head]...)
	w.followers[f] = true // The head and handoff registration share append's lock.
	w.mu.Unlock()
	go func() {
		defer close(f.done)
		defer target.Close(1008, "recorded handoff ended")
		send := func(entry recordedEntry) bool {
			select {
			case <-f.stop:
				return false
			default:
			}
			if err := target.Send(entry.path, entry.message); err != nil {
				return false
			}
			f.sent <- entry.sequence
			return true
		}
		for i, entry := range history {
			if !send(entry) {
				return
			}
			if i == 0 && pause {
				close(f.paused)
				select {
				case <-f.resume:
				case <-f.stop:
					return
				}
			}
		}
		for {
			select {
			case <-f.stop:
				return
			case entry := <-f.live:
				if !send(entry) {
					return
				}
			}
		}
	}()
	return head, f
}

// An asynchronous fixture root: selection and mounting use the production
// implementation, while this root supplies deterministic queued delivery. A
// forward hop receives and sends the same opaque Message to a second root.
type recordedRoot struct {
	mu        sync.Mutex
	receivers map[string]duplex.Receiver
	queue     chan recordedEntry
	stop      chan struct{}
	done      chan struct{}
	once      sync.Once
}

func newRecordedRoot() *recordedRoot {
	r := &recordedRoot{receivers: map[string]duplex.Receiver{}, queue: make(chan recordedEntry, 16), stop: make(chan struct{}), done: make(chan struct{})}
	go func() {
		defer close(r.done)
		for {
			select {
			case <-r.stop:
				return
			case entry := <-r.queue:
				key, _ := duplex.EncodePath(entry.path)
				r.mu.Lock()
				receiver := r.receivers[key]
				r.mu.Unlock()
				if receiver.Message != nil {
					receiver.Message(append([]string{}, entry.path...), entry.message)
				}
			}
		}
	}()
	return r
}
func (r *recordedRoot) Send(path []string, message duplex.Message) error {
	select {
	case <-r.stop:
		return duplex.ErrClosed
	default:
	}
	select {
	case r.queue <- recordedEntry{path: append([]string{}, path...), message: message}:
		return nil
	default:
		return fmt.Errorf("witness output queue full")
	}
}
func (r *recordedRoot) Receive(path []string, receiver duplex.Receiver) (func(), error) {
	key, err := duplex.EncodePath(path)
	if err != nil {
		return nil, err
	}
	r.mu.Lock()
	if _, exists := r.receivers[key]; exists {
		r.mu.Unlock()
		return nil, duplex.ErrReceiverExists
	}
	r.receivers[key] = receiver
	r.mu.Unlock()
	return func() { r.mu.Lock(); delete(r.receivers, key); r.mu.Unlock() }, nil
}
func (r *recordedRoot) Close(duplex.Code, string) error {
	r.once.Do(func() { close(r.stop) })
	<-r.done
	return nil
}

type recordedPresentation struct {
	wire   duplex.Wire
	root   *recordedRoot
	end    *recordedRoot
	values chan int
	closed chan int
	errors chan error
}

func presentRecorded(w *recordedWire) (*recordedPresentation, error) {
	p := &recordedPresentation{root: newRecordedRoot(), end: newRecordedRoot(), values: make(chan int, 32), closed: make(chan int, 4), errors: make(chan error, 4)}
	destination := duplex.At(duplex.Mount(map[string]duplex.Wire{"out": duplex.At(p.end, []string{"destination"})}), []string{"out"})
	p.wire = duplex.At(duplex.Mount(map[string]duplex.Wire{"outer": duplex.Mount(map[string]duplex.Wire{"in": duplex.At(p.root, []string{"source"})})}), []string{"outer", "in"})
	_, err := destination.Receive([]string{"tick"}, duplex.Receiver{Message: func(path []string, m duplex.Message) {
		if len(path) != 1 || path[0] != "tick" {
			p.errors <- fmt.Errorf("recorded destination received path %q", path)
			return
		}
		// A real application callback reenters the store. Calling this under
		// append exclusion deadlocks and fails the witness's deadline.
		_ = w.head()
		var value int
		if err := json.Unmarshal(m.Frame.Data, &value); err != nil {
			p.errors <- err
			return
		}
		p.values <- value
	}})
	if err == nil {
		_, err = p.wire.Receive([]string{"tick"}, duplex.Receiver{Message: func(path []string, message duplex.Message) {
			if err := destination.Send(path, message); err != nil {
				p.errors <- err
			}
		}, Closed: func(code duplex.Code, _ string) { _ = w.head(); p.closed <- int(code) }})
	}
	return p, err
}
func (p *recordedPresentation) close() {
	_ = p.wire.Close(1000, "done")
	_ = p.root.Close(1000, "done")
	_ = p.end.Close(1000, "done")
}
func recordedMessage(value int) duplex.Message {
	return duplex.Message{Frame: duplex.ProfileFrame{Version: 1, Kind: duplex.ProfileEvent, Data: json.RawMessage(fmt.Sprint(value))}}
}
func recordedWait[T any](ctx context.Context, channel <-chan T) (T, error) {
	select {
	case value := <-channel:
		return value, nil
	case <-ctx.Done():
		var zero T
		return zero, fail("timeout", "recorded wire witness: %v", ctx.Err())
	}
}
func recordedSent(ctx context.Context, f *recordedFollower, last int) error {
	for {
		value, err := recordedWait(ctx, f.sent)
		if err != nil {
			return err
		}
		if value == last {
			return nil
		}
	}
}
func (p *recordedPresentation) collect(ctx context.Context) ([]int, error) {
	// A fence through the same mount and forward hop proves every earlier
	// delivery has arrived; an extra replay cannot hide behind a length check.
	if err := p.wire.Send([]string{"tick"}, recordedMessage(0)); err != nil {
		return nil, err
	}
	values := []int{}
	for {
		select {
		case err := <-p.errors:
			return nil, err
		case value := <-p.values:
			if value == 0 {
				return values, nil
			}
			values = append(values, value)
		case <-ctx.Done():
			return nil, fail("timeout", "recorded delivery fence did not arrive")
		}
	}
}

func recordedHeadCase(ctx context.Context, before bool) (any, error) {
	w := newRecordedWire()
	defer w.Close(1000, "done")
	p, err := presentRecorded(w)
	if err != nil {
		return nil, err
	}
	defer p.close()
	source := duplex.At(duplex.Mount(map[string]duplex.Wire{"record": w}), []string{"record"})
	appendValue := func(value int) error { return source.Send([]string{"tick"}, recordedMessage(value)) }
	for value := 1; value <= 3; value++ {
		if err := appendValue(value); err != nil {
			return nil, err
		}
	}
	cut, next := "head_before_append", 4
	if before {
		cut, next = "append_before_head", 5
		if err := appendValue(4); err != nil {
			return nil, err
		}
	}
	head, f := w.attach(0, p.wire, 2, true)
	defer func() { f.once.Do(func() { close(f.stop) }); <-f.done }()
	if _, err := recordedWait(ctx, f.paused); err != nil {
		return nil, err
	}
	// Producer completion is observed while the replay writer is still held.
	produced := make(chan error, 1)
	go func() {
		for value := next; value <= 5; value++ {
			if err := appendValue(value); err != nil {
				produced <- err
				return
			}
		}
		produced <- nil
	}()
	if err, waitErr := recordedWait(ctx, produced); waitErr != nil {
		return nil, waitErr
	} else if err != nil {
		return nil, err
	}
	close(f.resume)
	if err := recordedSent(ctx, f, 5); err != nil {
		return nil, err
	}
	if err := appendValue(6); err != nil {
		return nil, err
	}
	if err := recordedSent(ctx, f, 6); err != nil {
		return nil, err
	}
	first, err := p.collect(ctx)
	if err != nil {
		return nil, err
	}
	second, err := presentRecorded(w)
	if err != nil {
		return nil, err
	}
	defer second.close()
	_, late := w.attach(3, second.wire, 2, false)
	defer func() { late.once.Do(func() { close(late.stop) }); <-late.done }()
	if err := recordedSent(ctx, late, 6); err != nil {
		return nil, err
	}
	after, err := second.collect(ctx)
	if err != nil {
		return nil, err
	}
	return map[string]any{"cut": cut, "head": head, "first": first, "after_three": after, "producer_progress": true, "callbacks_outside_append": true}, nil
}

func recordedStallCase(ctx context.Context) (any, error) {
	w := newRecordedWire()
	defer w.Close(1000, "done")
	for value := 1; value <= 3; value++ {
		_ = w.Send([]string{"tick"}, recordedMessage(value))
	}
	stalled, err := presentRecorded(w)
	if err != nil {
		return nil, err
	}
	defer stalled.close()
	healthy, err := presentRecorded(w)
	if err != nil {
		return nil, err
	}
	defer healthy.close()
	_, slow := w.attach(0, stalled.wire, 2, true)
	defer func() { slow.once.Do(func() { close(slow.stop) }); <-slow.done }()
	if _, err := recordedWait(ctx, slow.paused); err != nil {
		return nil, err
	}
	_, fast := w.attach(3, healthy.wire, 2, false)
	defer func() { fast.once.Do(func() { close(fast.stop) }); <-fast.done }()
	for value := 4; value <= 5; value++ {
		_ = w.Send([]string{"tick"}, recordedMessage(value))
		if err := recordedSent(ctx, fast, value); err != nil {
			return nil, err
		}
	}
	queued := len(slow.live)
	_ = w.Send([]string{"tick"}, recordedMessage(6))
	if err := recordedSent(ctx, fast, 6); err != nil {
		return nil, err
	}
	if _, err := recordedWait(ctx, slow.done); err != nil {
		return nil, err
	}
	code, err := recordedWait(ctx, stalled.closed)
	if err != nil {
		return nil, err
	}
	if err := stalled.wire.Send([]string{"tick"}, recordedMessage(99)); err == nil {
		return nil, fmt.Errorf("stalled carrier accepted after close")
	}
	underneath := make(chan int, 1)
	_, err = stalled.root.Receive([]string{"probe"}, duplex.Receiver{Message: func(_ []string, message duplex.Message) {
		var value int
		_ = json.Unmarshal(message.Frame.Data, &value)
		underneath <- value
	}})
	if err != nil {
		return nil, err
	}
	if err := stalled.root.Send([]string{"probe"}, recordedMessage(99)); err != nil {
		return nil, err
	}
	probe, err := recordedWait(ctx, underneath)
	if err != nil {
		return nil, err
	}
	_ = w.Send([]string{"tick"}, recordedMessage(7))
	if err := recordedSent(ctx, fast, 7); err != nil {
		return nil, err
	}
	values, err := healthy.collect(ctx)
	if err != nil {
		return nil, err
	}
	return map[string]any{"bound": 2, "queued_at_bound": queued, "closed": 1 + len(stalled.closed), "close_code": code, "healthy": values, "underneath": []int{probe}, "head": w.head(), "producer_progress": true}, nil
}

func (t *testee) recordedWireOps() map[string]func(request) (any, error) {
	return map[string]func(request) (any, error){"peer.recorded_wire_witness": func(r request) (any, error) {
		within, err := r.within()
		if err != nil {
			return nil, err
		}
		ctx, cancel := context.WithTimeout(context.Background(), within)
		defer cancel()
		cases := []any{}
		for _, before := range []bool{false, true} {
			observation, err := recordedHeadCase(ctx, before)
			if err != nil {
				return nil, err
			}
			cases = append(cases, observation)
		}
		stalled, err := recordedStallCase(ctx)
		if err != nil {
			return nil, err
		}
		return map[string]any{"cases": cases, "stalled": stalled}, nil
	}}
}

File: conformance/swift/NightseamConformance/Sources/NightseamTestee/RecordedWire.swift
import Bitwire
import Foundation
import NightseamDuplex
import NightseamRuntime

// A consumer append/follow composition of Wire. Sequence numbers are retained
// by this testee's store and never added to the profile's envelope.
private struct RecordedEntry: Sendable {
    let path: [String]
    let message: Message
    let sequence: Int
}

private struct RecordedFailure: Error, Sendable, CustomStringConvertible {
    let description: String
    init(_ description: String) { self.description = description }
}

private final class RecordedMailbox<Value: Sendable>: @unchecked Sendable {
    private let condition = NSCondition()
    private var values: [Value] = []
    var count: Int { condition.withLock { values.count } }
    func put(_ value: Value) {
        condition.lock(); values.append(value); condition.signal(); condition.unlock()
    }
    func take(until deadline: Date) throws -> Value {
        condition.lock(); defer { condition.unlock() }
        while values.isEmpty {
            guard condition.wait(until: deadline) else { throw RecordedFailure("Recorded wire witness timed out") }
        }
        return values.removeFirst()
    }
}

private final class RecordedFollower: @unchecked Sendable {
    let target: any Endpoint
    let bound: Int
    let sent = RecordedMailbox<Int>()
    let paused = RecordedMailbox<Bool>()
    let done = RecordedMailbox<Bool>()
    private let condition = NSCondition()
    private var live: [RecordedEntry] = []
    private var stopped = false
    private var resumed = false
    init(target: any Endpoint, bound: Int) { self.target = target; self.bound = bound }
    var queued: Int { condition.withLock { live.count } }
    func stop() { condition.lock(); stopped = true; condition.broadcast(); condition.unlock() }
    func resume() { condition.lock(); resumed = true; condition.broadcast(); condition.unlock() }
    func enqueue(_ entry: RecordedEntry) -> Bool {
        condition.lock(); defer { condition.unlock() }
        guard !stopped else { return false }
        guard live.count < bound else { stopped = true; condition.broadcast(); return false }
        live.append(entry); condition.signal(); return true
    }
    func run(history: [RecordedEntry], pause: Bool) {
        defer {
            try? target.close(code: 1008, reason: "Recorded handoff ended")
            done.put(true)
        }
        func send(_ entry: RecordedEntry) -> Bool {
            guard !condition.withLock({ stopped }) else { return false }
            do { try target.send(path: entry.path, message: entry.message) }
            catch { return false }
            sent.put(entry.sequence); return true
        }
        for (index, entry) in history.enumerated() {
            guard send(entry) else { return }
            if index == 0 && pause {
                paused.put(true)
                condition.lock()
                while !resumed && !stopped { condition.wait() }
                let stop = stopped
                condition.unlock()
                if stop { return }
            }
        }
        while true {
            condition.lock()
            while live.isEmpty && !stopped { condition.wait() }
            if stopped { condition.unlock(); return }
            let entry = live.removeFirst()
            condition.unlock()
            guard send(entry) else { return }
        }
    }
    func awaitSent(_ sequence: Int, until deadline: Date) throws {
        while try sent.take(until: deadline) != sequence {}
    }
}

private final class RecordedStore: Wire, @unchecked Sendable {
    private let lock = NSLock()
    private var entries: [RecordedEntry] = []
    private var followers: [UUID: RecordedFollower] = [:]
    var head: Int { lock.withLock { entries.count } }
    func send(path: [String], message: Message) throws {
        lock.withLock {
            let entry = RecordedEntry(path: path, message: message, sequence: entries.count + 1)
            entries.append(entry)
            let overflowed = followers.compactMap { id, follower in follower.enqueue(entry) ? nil : id }
            for id in overflowed { followers.removeValue(forKey: id) }
        }
    }
    func close(code: Int, reason: String) throws {
        let detached = lock.withLock { let copy = Array(followers.values); followers.removeAll(); return copy }
        for follower in detached { follower.stop() }
    }
    func attach(after: Int, target: any Endpoint, bound: Int, pause: Bool) -> (Int, RecordedFollower) {
        let follower = RecordedFollower(target: target, bound: bound)
        // Taking the head and installing the live follower share append's lock.
        let (head, history) = lock.withLock {
            let head = entries.count
            let history = Array(entries[after..<head])
            followers[UUID()] = follower
            return (head, history)
        }
        DispatchQueue.global().async { follower.run(history: history, pause: pause) }
        return (head, follower)
    }
}

// A deterministic asynchronous fixture origin. Selection and mount remain the
// production implementations; forwarding sends the same opaque Message through
// a second queued origin, exercising actual callbacks and relative paths.
private final class RecordedRoot: Endpoint, @unchecked Sendable {
    private let lock = NSLock()
    private let worker = DispatchQueue(label: "nightseam.recorded.root")
    private var receivers: [Data: Receiver] = [:]
    private var queue: [RecordedEntry] = []
    private var running = false
    private var ended = false
    func send(path: [String], message: Message) throws {
        let start = try lock.withLock {
            guard !ended else { throw RecordedFailure("Fixture root closed") }
            guard queue.count < 16 else { throw RecordedFailure("Fixture output queue full") }
            queue.append(RecordedEntry(path: path, message: message, sequence: 0))
            let start = !running; running = true; return start
        }
        if start { worker.async { self.run() } }
    }
    private func run() {
        while true {
            let next: (RecordedEntry, Receiver?)? = lock.withLock {
                guard !ended, !queue.isEmpty else { running = false; return nil }
                let entry = queue.removeFirst()
                return (entry, receivers[Data()])
            }
            guard let (entry, receiver) = next else { return }
            receiver?.message(entry.path, entry.message)
        }
    }
    func receive(receiver: Receiver) throws -> Detach {
        let key = Data()
        try lock.withLock {
            guard !ended else { throw RecordedFailure("Fixture root closed") }
            guard receivers[key] == nil else { throw RecordedFailure("Fixture receiver exists") }
            receivers[key] = receiver
        }
        return { [self] in _ = lock.withLock { receivers.removeValue(forKey: key) } }
    }
    func close(code: Int, reason: String) throws {
        lock.withLock { ended = true; queue.removeAll(); receivers.removeAll() }
    }
}

private func recordedMessage(_ value: Int) -> Message {
    Message(frame: ProfileFrame(kind: .event, data: Data(String(value).utf8)))
}

private final class RecordedPresentation: Sendable {
    let root: RecordedRoot
    let end: RecordedRoot
    let wire: any Endpoint
    let rootRoutes: Dispatcher
    let routes: [Dispatcher]
    let values: RecordedMailbox<Int>
    let closed: RecordedMailbox<Int>
    let errors: RecordedMailbox<String>
    let reentered: RecordedMailbox<Bool>
    init(store: RecordedStore) throws {
        let root = RecordedRoot(), end = RecordedRoot()
        let values = RecordedMailbox<Int>(), closed = RecordedMailbox<Int>()
        let errors = RecordedMailbox<String>(), reentered = RecordedMailbox<Bool>()
        let rootRoutes = try Dispatcher(root), endRoutes = try Dispatcher(end)
        let destinationRoutes = try Dispatcher(mount(["out": endRoutes.select(path: ["destination"])]))
        let sourceRoutes = try Dispatcher(mount(["outer": mount(["in": rootRoutes.select(path: ["source"])])]))
        let destination = destinationRoutes.select(path: ["out"])
        let wire = sourceRoutes.select(path: ["outer", "in"])
        self.rootRoutes = rootRoutes; self.routes = [endRoutes, destinationRoutes, sourceRoutes]
        _ = try destination.receive(receiver: Receiver(message: { path, message in
            guard Path.key(path) == Path.key(["tick"]) else { errors.put("Destination received wrong path"); return }
            // A real consumer callback takes append's lock. If delivery occurred
            // under that lock, the fence below would fail its bounded wait.
            _ = store.head
            reentered.put(true)
            guard let data = message.frame.data, let text = String(data: data, encoding: .utf8), let value = Int(text)
            else { errors.put("Destination received invalid event data"); return }
            values.put(value)
        }))
        _ = try wire.receive(receiver: Receiver(message: { path, message in
            do { try destination.send(path: path, message: message) }
            catch { errors.put(String(describing: error)) }
        }, closed: { code, _ in _ = store.head; closed.put(code) }))
        self.root = root; self.end = end; self.wire = wire
        self.values = values; self.closed = closed; self.errors = errors; self.reentered = reentered
    }
    func close() {
        try? wire.close(code: 1000, reason: "Done")
        rootRoutes.close(); for route in routes { route.close() }
        try? root.close(code: 1000, reason: "Done")
        try? end.close(code: 1000, reason: "Done")
    }
    func collect(until deadline: Date) throws -> [Int] {
        // The fence takes the same mounted forwarding route as all data. Thus
        // an extra replay cannot hide after collecting an expected count.
        try wire.send(path: ["tick"], message: recordedMessage(0))
        var result: [Int] = []
        while true {
            if errors.count > 0 { throw RecordedFailure(try errors.take(until: deadline)) }
            let value = try values.take(until: deadline)
            if value == 0 { return result }
            result.append(value)
        }
    }
}

private func recordedArray(_ values: [Int]) -> JSONValue { .array(values.map { .number(String($0)) }) }

private func recordedHeadCase(appendBeforeHead: Bool, deadline: Date) throws -> JSONValue {
    let store = RecordedStore()
    defer { try? store.close(code: 1000, reason: "Done") }
    let presentation = try RecordedPresentation(store: store)
    defer { presentation.close() }
    let source = at(store, path: [])
    let append: @Sendable (Int) throws -> Void = { value in
        try source.send(path: ["tick"], message: recordedMessage(value))
    }
    for value in 1...3 { try append(value) }
    if appendBeforeHead { try append(4) }
    let (head, follower) = store.attach(after: 0, target: presentation.wire, bound: 2, pause: true)
    defer { follower.stop() }
    _ = try follower.paused.take(until: deadline)
    let produced = RecordedMailbox<Result<Bool, RecordedFailure>>()
    DispatchQueue.global().async {
        do {
            for value in (appendBeforeHead ? 5 : 4)...5 { try append(value) }
            produced.put(.success(true))
        } catch { produced.put(.failure(RecordedFailure(String(describing: error)))) }
    }
    let progress = try produced.take(until: deadline).get()
    follower.resume()
    try follower.awaitSent(5, until: deadline)
    try append(6)
    try follower.awaitSent(6, until: deadline)
    let first = try presentation.collect(until: deadline)
    let second = try RecordedPresentation(store: store)
    defer { second.close() }
    let (_, late) = store.attach(after: 3, target: second.wire, bound: 2, pause: false)
    defer { late.stop() }
    try late.awaitSent(6, until: deadline)
    let after = try second.collect(until: deadline)
    return .object([
        "cut": .string(appendBeforeHead ? "append_before_head" : "head_before_append"),
        "head": .number(String(head)), "first": recordedArray(first), "after_three": recordedArray(after),
        "producer_progress": .bool(progress), "callbacks_outside_append": .bool(presentation.reentered.count > 0 && second.reentered.count > 0)
    ])
}

private func recordedStallCase(deadline: Date) throws -> JSONValue {
    let store = RecordedStore()
    defer { try? store.close(code: 1000, reason: "Done") }
    for value in 1...3 { try store.send(path: ["tick"], message: recordedMessage(value)) }
    let stalled = try RecordedPresentation(store: store), healthy = try RecordedPresentation(store: store)
    defer { stalled.close(); healthy.close() }
    let (_, slow) = store.attach(after: 0, target: stalled.wire, bound: 2, pause: true)
    defer { slow.stop() }
    _ = try slow.paused.take(until: deadline)
    let (_, fast) = store.attach(after: 3, target: healthy.wire, bound: 2, pause: false)
    defer { fast.stop() }
    for value in 4...5 {
        try store.send(path: ["tick"], message: recordedMessage(value))
        try fast.awaitSent(value, until: deadline)
    }
    let queued = slow.queued
    try store.send(path: ["tick"], message: recordedMessage(6))
    try fast.awaitSent(6, until: deadline)
    _ = try slow.done.take(until: deadline)
    let code = try stalled.closed.take(until: deadline)
    var rejected = false
    do { try stalled.wire.send(path: ["tick"], message: recordedMessage(99)) }
    catch { rejected = true }
    guard rejected else { throw RecordedFailure("Stalled carrier admitted after close") }
    let underneath = RecordedMailbox<Int>()
    _ = try stalled.rootRoutes.register(path: ["probe"], receiver: Receiver(message: { _, message in
        if let data = message.frame.data, let text = String(data: data, encoding: .utf8), let value = Int(text) { underneath.put(value) }
    }))
    try stalled.root.send(path: ["probe"], message: recordedMessage(99))
    let probe = try underneath.take(until: deadline)
    try store.send(path: ["tick"], message: recordedMessage(7))
    try fast.awaitSent(7, until: deadline)
    let values = try healthy.collect(until: deadline)
    return .object([
        "bound": .number(String(slow.bound)), "queued_at_bound": .number(String(queued)),
        "closed": .number(String(1 + stalled.closed.count)), "close_code": .number(String(code)),
        "healthy": recordedArray(values), "underneath": recordedArray([probe]),
        "head": .number(String(store.head)), "producer_progress": .bool(values.last == store.head)
    ])
}

func recordedWireWitness(withinMilliseconds: Int = 5_000) throws -> JSONValue {
    let deadline = Date().addingTimeInterval(Double(withinMilliseconds) / 1000)
    return .object([
        "cases": .array(try [false, true].map { try recordedHeadCase(appendBeforeHead: $0, deadline: deadline) }),
        "stalled": try recordedStallCase(deadline: deadline)
    ])
}
